server/util/grpc_server/grpc_server.go

```diff
@@ -68,112 +68,81 @@ func MaxRecvMsgSizeBytes() int {
 	return *gRPCMaxRecvMsgSizeBytes
 }
 
-type RegisterServices func(server *grpc.Server, env *real_environment.RealEnv)
+type Builder struct {
+	env  environment.Env
+	port int
+	ssl  bool
 
-func RegisterGRPCServer(env *real_environment.RealEnv, regServices RegisterServices) error {
-	grpcServer, err := newGRPCServer(env, *gRPCPort, nil, regServices)
-	if err != nil {
-		return err
-	}
-	env.SetGRPCServer(grpcServer)
-	if *gRPCOverHTTPPortEnabled {
-		env.SetMux(&gRPCMux{
-			env.GetMux(),
-			grpcServer,
-		})
-	}
-	return nil
+	hostPort string
+	server   *grpc.Server
 }
 
-func RegisterGRPCSServer(env *real_environment.RealEnv, regServices RegisterServices) error {
-	if !env.GetSSLService().IsEnabled() {
-		return nil
-	}
-	creds, err := env.GetSSLService().GetGRPCSTLSCreds()
-	if err != nil {
-		return status.InternalErrorf("Error getting SSL creds: %s", err)
-	}
-	grpcsServer, err := newGRPCServer(env, *gRPCSPort, grpc.Creds(creds), regServices)
-	if err != nil {
-		return err
-	}
-	env.SetGRPCSServer(grpcsServer)
-	return nil
-}
-
-func RegisterInternalGRPCServer(env *real_environment.RealEnv, regServices RegisterServices) error {
-	if *internalGRPCPort == 0 {
-		return nil
-	}
-
-	grpcServer, err := newGRPCServer(env, *internalGRPCPort, nil, regServices)
-	if err != nil {
-		return err
-	}
-	env.SetInternalGRPCServer(grpcServer)
-	return nil
+func (b *Builder) GetServer() *grpc.Server {
+	return b.server
 }
 
-func RegisterInternalGRPCSServer(env *real_environment.RealEnv, regServices RegisterServices) error {
-	if *internalGRPCSPort == 0 {
-		return nil
+func NewBuilder(env environment.Env, port int, ssl bool) (*Builder, error) {
```

Comment:
I was a little thrown off by the naming as this is not a "builder" in the conventional sense. What do you think instead of returning a `*Builder` you return `(*grpc.Server server, func() error, error)` and get rid of the `*Builder` struct altogether? You could also rename `NewBuilder` to `New` so at the call site the call becomes `s, start, err := grpc_server.New(...)`.

Reply:
Yeah, not a great name. In a previous iteration it was more builder-y. I'd prefer to keep the struct because in the future I think there'll be more configuration options added and I think it's a bit more clear that passing callbacks around, so I renamed this to GRPCServer.

Reply:
Do you have a specific example in mind of what you think will be added to the struct in the future? I'm not a big fan of making speculative changes for something that may not ever come to be. The struct can always be added in the future, as needed if more complexity is required. 

As far as returning a function, it's not that uncommon to do in Go code. I don't see a big difference between asking the user to call Start on the struct or calling the returned start function.

Reply:
I'm going to add a way to add additional RPC interceptors for cache proxy auth.

I know it's not uncommon to return callbacks in go and it's effectively the same thing, but I always find passing functions around more confusing than working with an object. I can change it if you'd really like but part of my motivation with this change was to remove the closures.